### PR TITLE
Storm 🌪️: Preload logo via <link rel="preload">

### PR DESCRIPTION
**Pull request by philipp-spiess**
Digging through the resource water-falling of our navbar component, I noticed that we start to load the logo relatively late, even though it's shown on any route. Additionally, we use the same SVG as the favicon but with a slightly different URL causing the browser to have to make another request:

Screenshot 2023-02-20 at 15 45 11

This PR adds a way for the server-generated HTML to include `<link rel="preload">` elements to preload the .svg in question.

It also fixes the favicon to now use the same URL. It does seem that `<link rel="preload">` is not used by the browser for the favicon but the request might still be cached with the right cache headers (and if that also doesn't apply, at least a CDN might be able to do that!)

The lazy loading nature of the image also caused it to flicker during page loads (this video is from dotcom):

https://user-images.githubusercontent.com/458591/220139052-355448d7-3a7a-4d8e-bb78-ea61dca7b8dc.mov

With preloading, it appears instantly. 💥 

## Future improvement

- [ ] We should also do this for the avatar if the user is logged in

### Before on localhost

Screenshot 2023-02-20 at 15 49 04

### After on localhost

Screenshot 2023-02-20 at 15 39 30
Screenshot 2023-02-20 at 15 51 14

### CODE REVIEW

File: cmd/frontend/internal/app/ui/app.html [hunk 1]
```diff
@@ -31,6 +31,8 @@
 	{{if .Manifest.AppCSSBundlePath}}<link rel="stylesheet" href="{{.Manifest.AppCSSBundlePath}}">{{end}}
 	<link id='sourcegraph-chrome-webstore-item' rel="chrome-webstore-item" href="https://chrome.google.com/webstore/detail/dgjhfomjieaadpoljlnidmbgkdffpack">
 	<link rel="search" href="/opensearch.xml" type="application/opensearchdescription+xml" title="Sourcegraph Search">
+	<link rel="preload" href="/.assets/img/sourcegraph-mark.svg?v2" as="image" />
```
Comment (philipp-spiess) [inline cmd/frontend/internal/app/ui/app.html:34]: This should use `assetsutil.URL`

File: cmd/frontend/internal/app/misc_handlers.go [hunk 2]
```diff
@@ -53,9 +54,9 @@ func favicon(w http.ResponseWriter, r *http.Request) {
 
 	// Add query parameter for cache busting.
 	query := url.Query()
-	query.Set("v", "2")
+	query.Set("v2", "")
 	url.RawQuery = query.Encode()
-	path := url.String()
+	path := strings.Replace(url.String(), "v2=", "v2", 1)
```
Comment (philipp-spiess) [inline cmd/frontend/internal/app/misc_handlers.go:59]: The URL we used for the favicon was not matching the URL used on the layout causing a second request for the same resource. This fixed this `?v=2` vs `?v2`

**Review by eseliger — APPROVED** [review]

**Review by mrnugget — APPROVED** [review]
Full stack engineering, baby!

File: cmd/frontend/internal/app/ui/handlers.go [hunk 3]
```diff
@@ -151,6 +161,10 @@ func newCommon(w http.ResponseWriter, r *http.Request, db database.DB, title str
 		Context:  jscontext.NewJSContextFromRequest(r, db),
 		Title:    title,
 		Manifest: manifest,
+		PreloadedAssets: &[]PreloadedAsset{
+			// sourcegraph-mark.svg is always loaded as part of the layout component
+			{As: "image", Href: assetsutil.URL("/img/sourcegraph-mark.svg").Path + "?v2"},
```
Comment (keegancsmith) [inline cmd/frontend/internal/app/ui/handlers.go:166]: Is it intentional to use Path here? Will this image href always be on the same host?

**Review by valerybugakov — APPROVED** [review]
Great improvement! Left a comment about potential logo overrides by the side config.

File: cmd/frontend/internal/app/ui/handlers.go (hunk 3 above)
Comment (philipp-spiess) [reply cmd/frontend/internal/app/ui/handlers.go:166]: @valerybugakov We don't know the dark mode setting on the backend which is always needed for brand-defined images

```
    const customBrandingLogoUrl = branding?.[themeProperty]?.[variant]
```

So for the case where a custom branding is defined, we can't statically generate the right image to preload. I can only skip it if that's the case I guess.

Comment (valerybugakov) [reply cmd/frontend/internal/app/ui/handlers.go]: @philipp-spiess, sounds good!

**Pull request merged**